core/src/main/java/io/fabric8/maven/core/service/BuildService.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2016 Red Hat, Inc.
+ *
+ * Red Hat licenses this file to you under the Apache License, version
+ * 2.0 (the "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package io.fabric8.maven.core.service;
+
+import java.io.File;
+
+import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
+import io.fabric8.maven.core.config.BuildRecreateMode;
+import io.fabric8.maven.core.config.OpenShiftBuildStrategy;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.docker.util.MojoParameters;
+import io.fabric8.maven.docker.util.Task;
+
+import org.apache.maven.plugins.annotations.Parameter;
+
+/**
+ * @author nicola
+ * @since 17/02/2017
+ */
+public interface BuildService {
+
+    /**
+     * Builds the given image using the specified configuration.
+     *
+     * @param config the build configuration
+     * @param imageConfig the image to build
+     */
+    void build(BuildServiceConfig config, ImageConfiguration imageConfig) throws Fabric8ServiceException;
```

Comment:
Hmm. `ImageConfiguration` is of course also part of d-m-p. It doesn't make sense to abstract that away as its central for the image definition, regardless whether we are doing Docker or OpenShift builds. That's definitely something that d-m-p and f-m-p are going to share, so the best solution is to extract that into a separate project (like `fabric8io/docker-model` or `fabric8io/container-model` to make it more generic). Or at least into a separate module in `d-m-p` (which is currently a single module, including the mojos).

Again something to think about, not in the scope of this PR.